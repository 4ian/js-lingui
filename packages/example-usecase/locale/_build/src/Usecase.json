{
  "This isn't translated": {
    "origin": [
      [
        "src/Usecase.js",
        28
      ]
    ]
  },
  "msg.label": {
    "defaults": "",
    "origin": [
      [
        "src/Usecase.js",
        32
      ]
    ]
  },
  "Hello World": {
    "origin": [
      [
        "src/Usecase.js",
        36
      ]
    ]
  },
  "My name is {name}": {
    "origin": [
      [
        "src/Usecase.js",
        40
      ]
    ]
  },
  "Read <0>more</0>.": {
    "origin": [
      [
        "src/Usecase.js",
        44
      ]
    ]
  },
  "{genderOfHost, select, female {{numGuests, plural, offset:1 =0 {{host} does not give a party.} =1 {{host} invites {guest} to her party.} =2 {{host} invites {guest} and one other person to her party.} other {{host} invites {guest} and # other people to her party.}}} male {{numGuests, plural, offset:1 =0 {{host} does not give a party.} =1 {{host} invites {guest} to his party.} =2 {{host} invites {guest} and one other person to her party.} other {{host} invites {guest} and # other people to her party.}}} other {{numGuests, plural, offset:1 =0 {{host} does not give a party.} =1 {{host} invites {guest} to their party.} =2 {{host} invites {guest} and one other person to her party.} other {{host} invites {guest} and # other people to her party.}}}}": {
    "origin": [
      [
        "src/Usecase.js",
        50
      ]
    ]
  }
}