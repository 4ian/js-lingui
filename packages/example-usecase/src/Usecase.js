// @flow
import React from 'react'
import { Trans, I18nProvider, Select, Plural } from 'lingui-react'

class NeverUpdate extends React.Component {
  props: {
    children?: any
  }

  shouldComponentUpdate () {
    return false
  }

  render () {
    return <div>{this.props.children}</div>
  }
}

class Usecase extends React.Component {
  props: {
    language: string,
    messages: Object,
    name?: string,
    genderOfHost?: 'male' | 'female',
    numGuests?: number,
    host?: string,
    guest?: string
  }

  render () {
    const {
      language = '', messages = {},
      name = 'Mononoke',
      genderOfHost = 'female',
      numGuests = 4,
      host = 'Wilma',
      guest = 'Fred'
    } = this.props

    return (
      <I18nProvider language={language} messages={messages}>
        <NeverUpdate>
          <ul>
            <li>
              <Trans className="untranslated">{"This isn't translated"}</Trans>
            </li>

            <li>
              <Trans className="customId" id="msg.label" />
            </li>

            <li>
              <Trans className="translated">Hello World</Trans>
            </li>

            <li>
              <Trans className="variable">My name is {name}</Trans>
            </li>

            <li>
              <Trans className="components">
                Read <a href="/mononoke">more</a>.
              </Trans>
            </li>

            <li>
              <Trans className="plural">
                <Select
                  value={genderOfHost}
                  female={
                    <Plural
                      value={numGuests}
                      offset="1"
                      _0={`${host} does not give a party.`}
                      _1={`${host} invites ${guest} to her party.`}
                      _2={`${host} invites ${guest} and one other person to her party.`}
                      other={`${host} invites ${guest} and # other people to her party.`}
                    />
                  }
                  male={
                    <Plural
                      value={numGuests}
                      offset="1"
                      _0={`${host} does not give a party.`}
                      _1={`${host} invites ${guest} to his party.`}
                      _2={`${host} invites ${guest} and one other person to her party.`}
                      other={`${host} invites ${guest} and # other people to her party.`}
                    />
                  }
                  other={
                    <Plural
                      value={numGuests}
                      offset="1"
                      _0={`${host} does not give a party.`}
                      _1={`${host} invites ${guest} to their party.`}
                      _2={`${host} invites ${guest} and one other person to her party.`}
                      other={`${host} invites ${guest} and # other people to her party.`}
                    />
                  }
                />
              </Trans>
            </li>
          </ul>
        </NeverUpdate>
      </I18nProvider>
    )
  }
}

export default Usecase
