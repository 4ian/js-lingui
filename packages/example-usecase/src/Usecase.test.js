// @flow
import fs from 'fs'
import path from 'path'
import React from 'react'
import { mount } from 'enzyme'

import Usecase from './Usecase'

const rmdir = (dir) => {
  const list = fs.readdirSync(dir)

  for (let i = 0; i < list.length; i++) {
    const filename = path.join(dir, list[i])
    const stat = fs.statSync(filename)

    if (filename === '.' || filename === '..') {
      // pass these files
    } else if (stat.isDirectory()) {
      // rmdir recursively
      rmdir(filename)
    } else {
      // rm fiilename
      fs.unlinkSync(filename)
    }
  }
  fs.rmdirSync(dir)
}

describe('example-usecase', function () {
  const messages = {}

  beforeAll(function () {
    messages.cs = JSON.parse(fs.readFileSync('./packages/example-usecase/locale/cs/messages.json').toString())
    messages.fr = JSON.parse(fs.readFileSync('./packages/example-usecase/locale/fr/messages.json').toString())
  })

  afterAll(function () {
    rmdir('./locale')
  })

  const getText = (element, props = {}) => {
    return mount(<Usecase {...props} language="cs" messages={messages} />).find(element).text()
  }

  const getHtml = (element, props = {}) => {
    return mount(<Usecase {...props} language="cs" messages={messages} />).find(element).html()
  }

  it('should render', function () {
    expect(mount(<Usecase language="cs" messages={messages} />)).toMatchSnapshot()
  })

  it('should render defaults with warning for untranslated', function () {
    expect(getText('.untranslated')).toEqual("This isn't translated")
  })

  it('should support custom message id', function () {
    expect(getText('.customId')).toEqual('Nápis')
  })

  it('should render translated string', function () {
    expect(getText('.translated')).toEqual('Ahoj světe')
  })

  it('should support variable substitution', function () {
    expect(getText('.variable')).toEqual('Jmenuji se Mononoke')
    expect(getText('.variable', { name: 'Fred' })).toEqual('Jmenuji se Fred')
  })

  it('should support nested elements', function () {
    expect(getHtml('.components')).toMatchSnapshot()
  })

  it('should support pluralization', function () {
    expect(getText('.plural'))
      .toEqual('Wilma invites Fred and 3 other people to her party.')
  })

  it('should update translation when language changes', function () {
    const node = mount(<Usecase language='cs' messages={messages} />)
    expect(node.find('.translated').text()).toEqual('Ahoj světe')

    node.setProps({ language: 'fr' })
    expect(node.find('.translated').text()).toEqual('Salut le monde!')
  })
})
