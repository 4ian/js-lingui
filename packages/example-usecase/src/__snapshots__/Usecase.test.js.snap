exports[`example-usecase should render 1`] = `
<Component
  language="cs"
  messages={
    Object {
      "cs": Object {
        "Full content of {articleName}": "Celé znění článku {articleName}",
        "Hello World": "Ahoj světe",
        "My name is {name}": "Jmenuji se {name}",
        "Read <0>more</0>": "Číst <0>dále</0>",
        "msg.label": "Nápis",
      },
      "fr": Object {
        "Hello World": "Salut le monde!",
        "My name is {name}": "Je m\'appelle {name}",
        "msg.label": "Label",
      },
    }
  }>
  <I18nProvider
    language="cs"
    messages={
      Object {
        "cs": Object {
          "Full content of {articleName}": "Celé znění článku {articleName}",
          "Hello World": "Ahoj světe",
          "My name is {name}": "Jmenuji se {name}",
          "Read <0>more</0>": "Číst <0>dále</0>",
          "msg.label": "Nápis",
        },
        "fr": Object {
          "Hello World": "Salut le monde!",
          "My name is {name}": "Je m\'appelle {name}",
          "msg.label": "Label",
        },
      }
    }>
    <NeverUpdate>
      <div>
        <UsecaseView>
          <div>
            <ul>
              <li>
                <WithI18n
                  className="untranslated"
                  id="This isn\'t translated">
                  <Trans
                    className="untranslated"
                    i18n={
                      I18n {
                        "_language": "cs",
                        "_messages": Object {
                          "cs": Object {
                            "Full content of {articleName}": "Celé znění článku {articleName}",
                            "Hello World": "Ahoj světe",
                            "My name is {name}": "Jmenuji se {name}",
                            "Read <0>more</0>": "Číst <0>dále</0>",
                            "msg.label": "Nápis",
                          },
                          "fr": Object {
                            "Hello World": "Salut le monde!",
                            "My name is {name}": "Je m\'appelle {name}",
                            "msg.label": "Label",
                          },
                        },
                        "plural": [Function],
                        "select": [Function],
                        "t": [Function],
                      }
                    }
                    id="This isn\'t translated">
                    <span
                      className="untranslated">
                      This isn\'t translated
                    </span>
                  </Trans>
                </WithI18n>
              </li>
              <li>
                <WithI18n
                  className="customId"
                  defaults=""
                  id="msg.label">
                  <Trans
                    className="customId"
                    defaults=""
                    i18n={
                      I18n {
                        "_language": "cs",
                        "_messages": Object {
                          "cs": Object {
                            "Full content of {articleName}": "Celé znění článku {articleName}",
                            "Hello World": "Ahoj světe",
                            "My name is {name}": "Jmenuji se {name}",
                            "Read <0>more</0>": "Číst <0>dále</0>",
                            "msg.label": "Nápis",
                          },
                          "fr": Object {
                            "Hello World": "Salut le monde!",
                            "My name is {name}": "Je m\'appelle {name}",
                            "msg.label": "Label",
                          },
                        },
                        "plural": [Function],
                        "select": [Function],
                        "t": [Function],
                      }
                    }
                    id="msg.label">
                    <span
                      className="customId">
                      Nápis
                    </span>
                  </Trans>
                </WithI18n>
              </li>
              <li>
                <WithI18n
                  className="translated"
                  id="Hello World">
                  <Trans
                    className="translated"
                    i18n={
                      I18n {
                        "_language": "cs",
                        "_messages": Object {
                          "cs": Object {
                            "Full content of {articleName}": "Celé znění článku {articleName}",
                            "Hello World": "Ahoj světe",
                            "My name is {name}": "Jmenuji se {name}",
                            "Read <0>more</0>": "Číst <0>dále</0>",
                            "msg.label": "Nápis",
                          },
                          "fr": Object {
                            "Hello World": "Salut le monde!",
                            "My name is {name}": "Je m\'appelle {name}",
                            "msg.label": "Label",
                          },
                        },
                        "plural": [Function],
                        "select": [Function],
                        "t": [Function],
                      }
                    }
                    id="Hello World">
                    <span
                      className="translated">
                      Ahoj světe
                    </span>
                  </Trans>
                </WithI18n>
              </li>
              <li>
                <WithI18n
                  className="variable"
                  id="My name is {name}"
                  params={
                    Object {
                      "name": "Mononoke",
                    }
                  }>
                  <Trans
                    className="variable"
                    i18n={
                      I18n {
                        "_language": "cs",
                        "_messages": Object {
                          "cs": Object {
                            "Full content of {articleName}": "Celé znění článku {articleName}",
                            "Hello World": "Ahoj světe",
                            "My name is {name}": "Jmenuji se {name}",
                            "Read <0>more</0>": "Číst <0>dále</0>",
                            "msg.label": "Nápis",
                          },
                          "fr": Object {
                            "Hello World": "Salut le monde!",
                            "My name is {name}": "Je m\'appelle {name}",
                            "msg.label": "Label",
                          },
                        },
                        "plural": [Function],
                        "select": [Function],
                        "t": [Function],
                      }
                    }
                    id="My name is {name}"
                    params={
                      Object {
                        "name": "Mononoke",
                      }
                    }>
                    <span
                      className="variable">
                      Jmenuji se Mononoke
                    </span>
                  </Trans>
                </WithI18n>
              </li>
              <li>
                <WithI18n
                  className="components"
                  components={
                    Array [
                      <a
                        href="/mononoke" />,
                    ]
                  }
                  id="Read <0>more</0>.">
                  <Trans
                    className="components"
                    components={
                      Array [
                        <a
                          href="/mononoke" />,
                      ]
                    }
                    i18n={
                      I18n {
                        "_language": "cs",
                        "_messages": Object {
                          "cs": Object {
                            "Full content of {articleName}": "Celé znění článku {articleName}",
                            "Hello World": "Ahoj světe",
                            "My name is {name}": "Jmenuji se {name}",
                            "Read <0>more</0>": "Číst <0>dále</0>",
                            "msg.label": "Nápis",
                          },
                          "fr": Object {
                            "Hello World": "Salut le monde!",
                            "My name is {name}": "Je m\'appelle {name}",
                            "msg.label": "Label",
                          },
                        },
                        "plural": [Function],
                        "select": [Function],
                        "t": [Function],
                      }
                    }
                    id="Read <0>more</0>.">
                    <span
                      className="components">
                      Read 
                      <a
                        href="/mononoke">
                        more
                      </a>
                      .
                    </span>
                  </Trans>
                </WithI18n>
              </li>
              <li>
                <WithI18n
                  className="plural"
                  id="{genderOfHost, select, female {{numGuests, plural, offset:1 =0 {{host} does not give a party.} =1 {{host} invites {guest} to her party.} =2 {{host} invites {guest} and one other person to her party.} other {{host} invites {guest} and # other people to her party.}}} male {{numGuests, plural, offset:1 =0 {{host} does not give a party.} =1 {{host} invites {guest} to his party.} =2 {{host} invites {guest} and one other person to her party.} other {{host} invites {guest} and # other people to her party.}}} other {{numGuests, plural, offset:1 =0 {{host} does not give a party.} =1 {{host} invites {guest} to their party.} =2 {{host} invites {guest} and one other person to her party.} other {{host} invites {guest} and # other people to her party.}}}}"
                  params={
                    Object {
                      "genderOfHost": "female",
                      "guest": "Fred",
                      "host": "Wilma",
                      "numGuests": 4,
                    }
                  }>
                  <Trans
                    className="plural"
                    i18n={
                      I18n {
                        "_language": "cs",
                        "_messages": Object {
                          "cs": Object {
                            "Full content of {articleName}": "Celé znění článku {articleName}",
                            "Hello World": "Ahoj světe",
                            "My name is {name}": "Jmenuji se {name}",
                            "Read <0>more</0>": "Číst <0>dále</0>",
                            "msg.label": "Nápis",
                          },
                          "fr": Object {
                            "Hello World": "Salut le monde!",
                            "My name is {name}": "Je m\'appelle {name}",
                            "msg.label": "Label",
                          },
                        },
                        "plural": [Function],
                        "select": [Function],
                        "t": [Function],
                      }
                    }
                    id="{genderOfHost, select, female {{numGuests, plural, offset:1 =0 {{host} does not give a party.} =1 {{host} invites {guest} to her party.} =2 {{host} invites {guest} and one other person to her party.} other {{host} invites {guest} and # other people to her party.}}} male {{numGuests, plural, offset:1 =0 {{host} does not give a party.} =1 {{host} invites {guest} to his party.} =2 {{host} invites {guest} and one other person to her party.} other {{host} invites {guest} and # other people to her party.}}} other {{numGuests, plural, offset:1 =0 {{host} does not give a party.} =1 {{host} invites {guest} to their party.} =2 {{host} invites {guest} and one other person to her party.} other {{host} invites {guest} and # other people to her party.}}}}"
                    params={
                      Object {
                        "genderOfHost": "female",
                        "guest": "Fred",
                        "host": "Wilma",
                        "numGuests": 4,
                      }
                    }>
                    <span
                      className="plural">
                      Wilma invites Fred and 3 other people to her party.
                    </span>
                  </Trans>
                </WithI18n>
              </li>
            </ul>
          </div>
        </UsecaseView>
      </div>
    </NeverUpdate>
  </I18nProvider>
</Component>
`;

exports[`example-usecase should support nested elements 1`] = `"<span class=\"components\"><!-- react-text: 14 -->Read <!-- /react-text --><a href=\"/mononoke\">more</a><!-- react-text: 16 -->.<!-- /react-text --></span>"`;
